slack/management/commands/fetch_slack_activity.py

```diff
@@ -0,0 +1,314 @@
+                "image_48": user_data["user"]["profile"].get("image_48", ""),
+            },
+        )
+        USERS_CACHE[user_id] = obj
+        return obj
+
+
+def should_track_message(message):
+    # These are not regular messages
+    # https://api.slack.com/events/message#subtypes
+    return message.get("subtype") in {None, "me_message"} and "bot_id" not in message
+
+
+def fill_channel_gap(gap, debug):
+    logger.info(
+        "Fetching channel history for %r (%r) in range (%s,%s)",
+        gap.channel.name,
+        gap.channel.id,
+        gap.oldest_message_ts,
+        gap.newest_message_ts,
+    )
+    pages = channel_messages_in_range(
+        channel=gap.channel.id,
+        latest=gap.newest_message_ts,
+        oldest=gap.oldest_message_ts,
+    )
+    first = True
+    for page in pages:
+        # use a separate transaction per page to allow restoring from an
+        # interrupted run.
+        with transaction.atomic():
+            for message in page:
+                if first and gap.newest_message_ts is None:
+                    gap.channel.last_update_ts = message["ts"]
+                    gap.channel.save()
+                    first = False
+                # Shrink the gap, but no need to save until we've finished this
+                # page (transactionally).
+                gap.newest_message_ts = message["ts"]
+
+                if not should_track_message(message):
+                    continue
+
+                if "user" in message:
+                    user = get_or_create_user(message["user"])
+                    if debug:
+                        gap.channel.seenmessage_set.create(ts=message["ts"])
+                    SlackActivityBucket.track_activity(gap.channel, user, message["ts"])
+
+                if message.get("thread_ts"):
+                    # Track this thread in the db to be able to check for
+                    # updates later.
+                    Thread.objects.create(
+                        channel=gap.channel,
+                        thread_ts=message["thread_ts"],
+                        # None indicates that this thread still must be updated
+                        # even if it's old.
+                        last_update_ts=None,
+                    )
+            gap.save()
+            logger.debug(
+                "Channel %r retrieved up to %s (%s)",
+                gap.channel.name,
+                # for the 'up to current' gap, newest_message_ts will be None
+                # and instead oldest_message_ts will be where we stopped.
+                gap.newest_message_ts or gap.oldest_message_ts,
+                parse_ts(gap.newest_message_ts or gap.oldest_message_ts),
+            )
+    # If we get here we must have gotten up to gap.oldest_message_ts, the gap
+    # is now empty. If we're interrupted before we get here, the gap will stay
+    # and be picked up from where we left off on the next run.
+    gap.delete()
+
+
+def do_thread(thread, debug):
```

Comment:
Please add a docstring to this function as well.

```suggestion
def do_thread(thread: Thread, debug: bool):
```